### PR TITLE
Add docs on how we style UI

### PR DESCRIPTION
**Pull request by felixfbecker**
This adds documentation for our current styling practices across web app, browser extension and shared UI and describes our unique constraints.

cc @rrhyne @AlicjaSuska

### CODE REVIEW

File: handbook/engineering/web/styling.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# Styling UI
+
+Sourcegraph has many UI components. A unique constraint for these is that they need to run withing different _environments_.
```
[the remark this answers is not in the document]
Comment (felixfbecker) [reply handbook/engineering/web/styling.md:3]: Thanks!

**Comment by rrhyne** [comment]
We will strive to mirror the same strategies where possible in Figma, especially these lines:

> The best approach is to not reference colors at all and use building blocks that have borders, text colors etc defined.
This saves code and makes it easy to maintain design consistency even if we want to change colors in the future.
When that is not possible (for example UI contributed by extensions), prefer to reference CSS variables with semantic colors like `var(--danger)`, `var(--success)`, `var(--border-color)`, `var(--body-bg)` etc.

File: handbook/engineering/web/styling.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+4. Support for advanced CSS features, like state selectors, pseudo elements, flexbox, grid, media queries, CSS variables, ...
+
+## Environments
+
+### Host-agnostic UI
+
+Components that need to run in different environments (any UI shared between our browser extension and the webapp) adopt styles from their environments through configurable CSS class names (as opposed to trying to replicate the styling with copied CSS).
+A component may accept multiple class names for different elements of the component.
+An example of this is `<HoverOverlay/>`: see how [the different props it accepts](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4047521a92904054e782d341001d08d61945c86f/-/blob/shared/src/hover/HoverOverlay.tsx?subtree=true#L27-39) for its child components' class names are pased in the [webapp](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4047521a92904054e782d341001d08d61945c86f/-/blob/web/src/components/shared.tsx?subtree=true#L35-44) and in [code host integrations](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4047521a92904054e782d341001d08d61945c86f/-/blob/browser/src/shared/code-hosts/shared/codeHost.tsx?subtree=true#L443:1).
```
Comment (felixfbecker) [inline handbook/engineering/web/styling.md:26]:
```suggestion
An example of this is `<HoverOverlay/>`: see how [the different props it accepts](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4047521a92904054e782d341001d08d61945c86f/-/blob/shared/src/hover/HoverOverlay.tsx#L27-39) for its child components' class names are passed in the [webapp](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4047521a92904054e782d341001d08d61945c86f/-/blob/web/src/components/shared.tsx#L35-44) and in [code host integrations](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4047521a92904054e782d341001d08d61945c86f/-/blob/browser/src/shared/code-hosts/shared/codeHost.tsx#L443:1).
```

**Comment by felixfbecker** [comment]
Friendly ping @marekweb @twop

**Comment by twop** [comment]
Thanks for writing this! I think it gives a really nice overview.

This may not be the feedback for the documentation itself (it is great as it is) but more on a meta level:
- as an engineer it is up to me to figure out which components do I need to use out of bootstrap and how that translates into actual react components. The doc shows the direction but not as useful as a prescription, which puts more responsibility on each engineer to "do the right thing" and "connect the dots". 
- there is no specific framework for composing smaller pieces into bigger pieces of UI. Especially when it comes to authoring react components (hint to component library).
- Layout is left as a guidance for engineers, again, which puts more responsibility on engineers to figure out which whether Flex box or CSS grid should be used, and how to solve a specific problem.
- I think it is worth building a shared vocabulary between designers and engineers when it comes to components and layouts, and it probably should operate on a higher level than css grid/flex box. I think It is extremely valuable for any handoff process or conversation to say a word which has a precise meaning for all participants (example: spacing-small -> 0.5rem, stack -> vertical layout etc).

If my thoughts have a ground then I worry about what we demand of engineers and designers to be productive in building new UIs.

Please treat the above not as feedback for this PR specifically, but more as my collection of thoughts on this topic (they don't need to be addressed or replied).

**Comment by felixfbecker** [comment]
@twop thanks for sharing! I'll merge this as docs on our current state and share my thoughts on your points when I find the time :)

**Pull request merged**